### DISCUSSION TITLE
Other games

### DISCUSSION THREAD

**Discussion by minecraftchest1** [openmod/openmod#514] [General]
I have just found this project, and I am wondering how to add support for other games such as minecraft, factorio, space engines, farming simular, etc.
I am a bit of a programmer, and I can probably help a bit.

**Comment by Trojaner** [comment]
Hello and welcome 👍 

Keep in mind that OpenMod currently only supports .NET based games, so it cannot support e.g. Minecraft. It works best on Unity games but can be basically used for any .NET based game. You could probably wrap native stuff to provide a plugin/mod API for games made in e.g. C or C++ such as Factorio, but this will be difficult if the game itself provides no API.

After creating the C# project ([here is an example .csproj](https://github.com/openmod/openmod/blob/main/unturned/OpenMod.Unturned/OpenMod.Unturned.csproj)), create a bootstrapper class that is somehow loaded or called by the game. Currently OpenMod has no own game patcher and relies on either the game loading the dlls natively or modloaders to load it. 

Use the following code to bootstrap OpenMod:
```cs
public class OpenModBootstrapper : IDisposable
{
    public Runtime.Runtime? OpenModRuntime { get; private set; }

    public void Initialize()
    {
        var assemblies = new List<Assembly>
        {
            typeof(OpenModBootstrapper ).Assembly,
        };

        OpenModRuntime = new Runtime.Runtime();
        OpenModRuntime.Init(assemblies, parameters);
    }

   public void Dispose()
   {
       OpenModRuntime?.Dispose();
       OpenModRuntime = null;
   }
}
```

Of course you can adjust this depending on your set up. In short, create a runtime and call `runtime.Init` on load and `runtime.Dispose` on unload / shutdown. It is important that the runtime is shutdown, otherwise data may not be saved.

After that you will have to implement the following services:
- `IOpenModHost`, initializes game specific stuff, e.g. console commands integration (for example, `FactorioOpenModHost`)
- `IGameHostInformation`, provides information such as server name, game, game version, port, etc. (for example, `FactorioGameHostInformation`)
- `IUser` (you can inherit it from UserBase), represents a player as OpenMod user (for example, `FactorioUser`)
- `IConsoleActorAccessor` (see [here](https://github.com/openmod/openmod/blob/main/standalone/OpenMod.Standalone/ConsoleActorAccessor.cs))

[Here](https://github.com/openmod/openmod/tree/main/standalone/OpenMod.Standalone) is a bare minimum implementation (but with no game related services).

After implementing these services you need to implement commands. To do this, hook the game's command and chat messages. Cancel native chat events so that commands don't show up in chat. Then call `ICommandExecutor.ExecuteAsync` to make OpenMod handle the command instead.

After doing this a basic implementation should be ready. After that you can start implementing our universal game API that allows to make universal plugins supported by all OpenMod games. This is optional and usually a long term task. The universal API provides abstractions over common games functions such as entities, NPCs, vehicles, etc. Universal APIs are defined [here](https://github.com/openmod/openmod/tree/main/extensions/OpenMod.Extensions.Games.Abstractions).

[Here](https://github.com/openmod/openmod/tree/main/unturned/OpenMod.Unturned) is an example full implementation of the universal API.

**Reply by minecraftchest1** [reply]
I figured that some proxy work would have to be done for non dotnet ganes. I'll take a look at the links you provided.
